### PR TITLE
Fix wrong connection close

### LINKED ISSUE #3829: Invalid connection close on post requests spanning multiple chunks
When post request contains more than two chunks over the network, we assume it is wrong and close the connection (it works, but we send `Connection: close`)

### PR DESCRIPTION
**Pull request by tomas-langer**
Resolves #3829 

Entity is consumed for any request that completed and did not subscribe (including JAX-RS).
We were consuming the entity anyway at a later stage (when we already did a connection close). For connection:close cases, entity is ignored if not consumed.

This keeps original handling and improved it even for requests with more than one chunk in request entity, while simplifying the code.

Signed-off-by: Tomas Langer <tomas.langer@oracle.com>

### CODE REVIEW

File: webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java [hunk 1]
```diff
@@ -195,23 +194,24 @@ public void writeStatusAndHeaders(Http.ResponseStatus status, Map<String, List<S
...
+                        entityRequested = false;
+                    }
+                }
+                if (entityRequested) {
+                    HttpRequest request = requestContext.request();
+                    LOGGER.warning("Entity was requested and not fully consumed before a response is sent. "
+                                           + "This is not supported. Connection will be closed. Please fix your route for "
+                            + request.method() + " " + request.uri());
+                }
+                if (!isWebSocketUpgrade) {
```
[the remark this answers is not in the document]
Comment (danielkec) [reply webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java:212]: Maybe moving the expression to the `RequestContext` would be best:
```java
    boolean partiallyDrainedStream() {
        return !requestCompleted()
                && !requestCancelled()
                && (hasRequests() || hasEmitted());
    }
```

**Review by danielkec — CHANGES_REQUESTED** [review]

File: webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (tomas-langer) [reply webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java]: Fixed

**Review by danielkec — APPROVED** [review]

File: webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java [hunk 2]
```diff
@@ -220,8 +239,8 @@ public void writeStatusAndHeaders(Http.ResponseStatus status, Map<String, List<S
 
         // Content length optimization attempt
         if (!lengthOptimization) {
-            LOGGER.fine(() -> log("Writing headers %s", status));
             requestEntityAnalyzed = requestEntityAnalyzed.thenApply(listener -> {
+                LOGGER.fine(() -> log("Writing headers %s", status));
```
Comment (spericas) [inline webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java:243]: Given the assignment in the line above, shouldn't we renamed `requestEntityAnalyzed`? Its name is too close to the new one added, yet its completion reflects more than just the request entity being processed?

Comment (tomas-langer) [reply webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java:243]: Not sure I understand - it is the same future. We just need to keep the original instance and complete on it, as the "last" isntance in the field will not reflect `thenApply` and other methods registered on previous instances.

Comment (spericas) [reply webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java:243]: That's exactly the point, the last instance on this field is related to the writing of the response and no longer the analysis of the request entity (as its name suggests). My comment is just about naming, not semantics. Certainly not a stopper.

**Review by spericas — APPROVED** [review]

**Pull request merged**